### ISSUE TITLE
Keyboard support via link port and emulation

### ISSUE THREAD

**Issue by Q-Bert-Reynolds** [chrismaltby/gb-studio#1617]
**Is your feature request related to a problem? Please describe.**
Text entry on a GameBoy is slow and playing music is hard. It would be a lot cooler if I could just plug a [PS/2 keyboard into the link port](https://youtu.be/qeRKhU0vp3c), use the [insideGadget's USB keyboard adapter](https://youtu.be/XhztoCdSls4), or leverage the power of emulators to expand the number of buttons available to the GameBoy. 

**Describe the solution you'd like**
Enable keyboard support on the settings page. More keybindings are available in the attach script to button event. A new scene in Sample Project that includes typing your character's name.

**Describe alternatives you've considered**
I made my own [scene editor and game engine](https://x.com/i/status/1360704152377122826) for GameBoy. [It has keyboard support](https://x.com/i/status/1402097579513483264), but I'm tired of building it, and it's not as easy to use as GB Studio.

**Additional context**
I've already written [a serial interface that supports PS/2 and USB keyboards](https://bitbucket.org/q_bert_reynolds/beisbolgb/src/master/src/keyboard/keyboard.asm). I just need to convert it for use with GBDK and add the editor bits and you'll be able to make GB Studio games that support keyboards on real hardware. What I'm less sure about is the emulation side. I'm thinking it will be a lot easier to emulate insideGadget's USB adapter since it sends 8 bit packets and the PS/2 keyboards send 11, but I haven't really looked into it.

**Comment by chrismaltby** [comment]
Very cool, this sounds like it should be an engine plugin rather than within the core itself (could add it to the official repo I'm working on https://github.com/gb-studio-dev/gb-studio-plugins). It might be possible to do this already without any core changes but if there are changes needed happy to help make this work

**Comment by Q-Bert-Reynolds** [comment]
I can see the serial interrupt as an engine plugin and the keybinding stuff as an events plugin, but I'm assuming I'd have to edit [this bit in GameBoyCore.js](https://github.com/chrismaltby/gb-studio/blob/660a55f593c64ed0d1b7a48b25ed9b813638bf66/appData/js-emulator/js/GameBoyCore.js#L5817C1-L5817C144) to get it to work in web builds.

**Comment by Q-Bert-Reynolds** [comment]
Holup. Are there two different emulators? One for builds one for the play button? I see Binjgb in `appData/wasm` and js-emulator in `appData`.

**Comment by pau-tomas** [comment]
Yeah. No reason to have the old emulator now. Binjgb is compiled to wasm but I wonder if a javascript wrapper could access the right memory addresses.

Maybe the way is to add support to Workboy to upstream Binjgb via PR. Similar to what Sameboy does [here](https://github.com/LIJI32/SameBoy/blob/master/Core/workboy.c).

**Comment by Q-Bert-Reynolds** [comment]
I may be a bit obsessed with the WorkBoy. Tried recreating it at one point. Got as far as a [3D model](https://i.redd.it/lhpbo0ztswv71.jpg), some kicad files, and a cursory read through of the disassembled WorkBoy rom. One day.

**Comment by maxoakland** [comment]
That's a fascinating idea
